internal/controller/clusterdeployment_controller.go

```diff
@@ -215,109 +210,104 @@ func (r *ClusterDeploymentReconciler) reconcileUpdate(ctx context.Context, cd *k
 }
 
 func (r *ClusterDeploymentReconciler) updateCluster(ctx context.Context, cd *kcm.ClusterDeployment, clusterTpl *kcm.ClusterTemplate) (ctrl.Result, error) {
-	l := ctrl.LoggerFrom(ctx)
-
 	if clusterTpl == nil {
 		return ctrl.Result{}, errors.New("cluster template cannot be nil")
 	}
 
+	l := ctrl.LoggerFrom(ctx)
+
+	r.initClusterConditions(cd)
+
 	if !clusterTpl.Status.Valid {
-		errMsg := "provided template is not marked as valid"
-		apimeta.SetStatusCondition(cd.GetConditions(), metav1.Condition{
-			Type:    kcm.TemplateReadyCondition,
-			Status:  metav1.ConditionFalse,
-			Reason:  kcm.FailedReason,
-			Message: errMsg,
-		})
-		return ctrl.Result{}, errors.New(errMsg)
+		err := fmt.Errorf("ClusterTemplate %s is not marked as valid", client.ObjectKeyFromObject(clusterTpl))
+		r.setCondition(cd, kcm.TemplateReadyCondition, err)
+		if r.IsDisabledValidationWH {
+			l.Error(err, "template is not valid, will not retrigger this error")
+			return ctrl.Result{}, nil
+		}
+		return ctrl.Result{}, err
 	}
+	r.setCondition(cd, kcm.TemplateReadyCondition, nil)
 	// template is ok, propagate data from it
 	cd.Status.KubernetesVersion = clusterTpl.Status.KubernetesVersion
 
-	apimeta.SetStatusCondition(cd.GetConditions(), metav1.Condition{
-		Type:    kcm.TemplateReadyCondition,
-		Status:  metav1.ConditionTrue,
-		Reason:  kcm.SucceededReason,
-		Message: "Template is valid",
-	})
+	var cred *kcm.Credential
+	if r.IsDisabledValidationWH {
+		l.Info("Validating ClusterDeployment upgrade sequence")
+		upgrades, err := r.getAvailableUpgrades(ctx, clusterTpl)
+		if err != nil {
+			l.Error(err, "failed to get available upgrades")
+			return ctrl.Result{}, err
+		}
+
+		var templateErr error
+		if !slices.Contains(upgrades, cd.Spec.Template) {
+			templateErr = fmt.Errorf("the upgrade sequence is not allowed: ClusterDeployment %s cannot be upgraded to %s", client.ObjectKeyFromObject(cd), cd.Spec.Template)
```

Comment:
I'm afraid the upgrade sequence validation will not work properly in the current implementation if the webhook is disabled. Right now, we don’t verify whether `spec.template` has changed. Instead, the controller always checks if the current template is available for an upgrade.
What happens if I haven’t configured any `ClusterTemplateChain`, and no upgrades are available? In that case, the controller will continuously fail, since the initial template isn’t present in the list of valid upgrade targets.
Or am I mistaken?

Reply:
Also, I guess we should add a watcher on ClusterTemplateChains as well (to catch the situation when the upgrade sequence for a particular template is updated). What do you think?

Reply:
Another concern is that we probably shouldn't validate k8s compatibility and other checks if the cluster is already deployed. What do you think?

Reply:
You are right about checking the new values rather than the old ones, it should be reworked. The problem is that I am not sure in which way: either we just drop this check or we validate a new (the current one) template reference against the `status` if it is not empty. The second option sounds preferable. If the latter is okay then the chains must be watched, that is correct 

> if the cluster is already deployed

The current implementation of the WH does not take this into account, should we also change the approach? What is the definition of the "already deployed Cluster"? Is it cluster with any `status` and without `dryRun` set, or is it one with the ready status?

Reply:
Removed upgrade sequence validations since they might be ambiguous and error-prone. Left the compatibility check as it is unclear how to determine whether a cluster has already been deployed and how to distinguish updates when it has already been deployed

Reply:
Discussed offline. Decided to drop the upgrade sequence validation from the reconciler.